lib/libgpanel/gpanel.3
.\" No copyright (2015) - Serge Vakulenko (serge@vak.ru)
.\"
.TH GPANEL 3 "June 10, 2023"
.UC 6
.SH NAME
gpanel_open, gpanel_close, gpanel_clear, gpanel_pixel, gpanel_line, gpanel_rect, gpanel_fill, gpanel_circle, gpanel_image, gpanel_char, gpanel_text, panel_text_width \- graphics panel routines
.SH SYNOPSIS
.nf
.PP
.ft B
#include <sys/gpanel.h>
.PP
.ft B
int gpanel_open(const char *devname);
.PP
.ft B
void gpanel_close(void);
.PP
.ft B
void gpanel_clear(int color, int *xsize, int *ysize);
.PP
.ft B
void gpanel_pixel(int color, int x, int y);
.PP
.ft B
void gpanel_line(int color, int x0, int y0, int x1, int y1);
.PP
.ft B
void gpanel_rect(int color, int x0, int y0, int x1, int y1);
.PP
.ft B
void gpanel_fill(int color, int x0, int y0, int x1, int y1);
.PP
.ft B
void gpanel_circle(int color, int x, int y, int radius);
.PP
.ft B
void gpanel_image(int x, int y, int width, int height,
                  const unsigned short *data);
.PP
.ft B
void gpanel_char(const struct gpanel_font_t *font, int color,
                 int background, int x, int y, int sym);
.PP
.ft B
void gpanel_text(const struct gpanel_font_t *font, int color,
                 int background, int x, int y, const char *text);
.PP
.ft B
int gpanel_text_width(const struct gpanel_font_t *font,
                      const char *text, int nchars);
.fi
.bp
.SH DESCRIPTION
This library provides a set of routines for access to a graphics panel
like TFT LCD or OLED displays.  HX8357 display on a Picadillo-35T board
is an example of such display. Use cc option
.I -lgpanel
to link the gpanel library.
.PP
.IR gpanel_open
\- This routine opens a connection to the display driver.
The argument to this function is the name of the device, like "/dev/tft0".
Negative value is returned in case of error.
.PP
.IR gpanel_close
\- Close a connection to the device.
.PP
.IR gpanel_clear
\- This routine is used to clear the whole screen to a given color,
optionally switch the display resolution and orientation,
and get the display size in pixels.
On input, parameters
.I xsize
and
.I ysize
should contain the desired dimensions of the display, or zeros to use
the default screen size and orientation. On output,
.I xsize
and
.I ysize
are set to the current display size.
.PP
.IR gpanel_pixel
\- Draw a single pixel of a specified color.
.PP
.IR gpanel_line
\- Draw a line.
.PP
.IR gpanel_rect
\- Draw a rectangular frame (not filled).
.PP
.IR gpanel_fill
\- Draw a rectangle filled with specified color.
.PP
.IR gpanel_circle
\- Draw a circle with a given center and radius.
.PP
.IR gpanel_image
\- Draw an arbitrary image: fill a rectangular area with pixel values
from a user-supplied data.
.PP
.IR gpanel_char
\- Draw a single character with a specified Unicode encoding.
.PP
.IR gpanel_text
\- Draw a text string in UTF-8 encoding.
.PP
.IR gpanel_text_width
\- Compute a width in pixels for a text string in UTF-8 encoding.
.PP
.SH EXAMPLE
.ft R
#include <stdio.h>
.br
#include <sys/gpanel.h>

int main()
.br
{
    int xsize, ysize;

    if (gpanel_open("/dev/tft0") < 0) {
        printf("Cannot open display\n");
        exit(-1);
    }
    /* Get screen size */
    gpanel_clear(0, &xsize, &ysize);

    /* Draw a white rectangle of max size */
    gpanel_rect(0xffff, 0, 0, xsize-1, ysize-1);
    return 0;
.br
}
.fi
.SH FONTS
.ft R
struct gpanel_font_t {
    const char *    name;         /* font name */
    int             maxwidth;     /* max width in pixels */
    unsigned int    height;       /* height in pixels */
    int             ascent;       /* ascent (baseline) height */
    int             firstchar;    /* first character in bitmap */
    int             size;         /* font size in characters */
    const unsigned short *bits;   /* 16-bit right-padded bitmap data */
    const unsigned short *offset; /* offsets into bitmap data */
    const unsigned char *width;   /* character widths or 0 if fixed */
    int             defaultchar;  /* default char (not glyph index) */
    long            bits_size;    /* number of words of bits */
.br
};
.PP
Fonts are specified via a data structure
.I struct\ gpanel_font_t
, which defines character sizes and glyph images.
Font data for some commonly used fonts are available in
.I /usr/share/examples/gpanel/fonts
directory.
A utility
.I convbdf
can be used to convert XWindows BDF fonts to a struct gpanel_font_t data.
When combined with
.I otf2bdf
utility, this allows to generate font data for any TrueType font,
like Verdana or Lucida Sans Unicode.
.SH BUGS
Data and font pointers in gpanel_image and gpanel_text routines
are not checked for validity, and can cause unexpected errors when not correct.
.SH FILES
.ta \w'/usr/share/examples/gpanel  'u
/usr/share/examples/gpanel  Examples of using the \fIlibgpanel\fP library.
.PP
/dev/tft0                   Device name for hx8357 TFT display driver.


ChangeLog
===========================================================================
2020-10-20
Tuesday, October 20, 2020

Started a ChangeLog file to keep track of my daily progress. I will need to
add in issues I have previously come up with and that which are not yet
included here.

I pushed a commit today of GPIO toggling (for the onboard LEDs) which goes
into great detail of each required part.
There are four registers to configure, in addition to configuring the RCC
clock. There are two ways to set or reset a GPIO pin once it is all set up:
by a read/modify/write dance, or with an atomic bit set or reset operation.

For setting up the GPIO (for port D, for the LEDs), this needs to be done:
 Set up the gpio:
1. clock through the RCC_AHB1ENR register.
2. mode through the GPIOD_MODER register.
3. output as push-pull or open-drain through the GPIOD_OTYPER register.
4. port speed through the GPIOD_OSPEEDR register.
5. pull-up or pull-down resistors through the GPIOD_PUPDR register.

For actually using the GPIO as an output there are two ways:
1. Read/modify/write of the GPIOD_ODR Output Data Register
2. Writing to the GPIOD_BSRR Bit Set/Reset Register

Whichever one is used (or available) is chip vendor dependent. These are
both available for STM32F4 chips.

I put the linker script in the testing base directory (with Makefile.inc)
so that all testing programs can use it, instead of having each program
duplicating the linker script.

The program I finished today has links to all the documentation I have used
to make that program. I will end up putting them in this repository.

I added a .gitignore to the testing directory. Not sure if it is best to
use a global one yet, so just this local one for now.

===========================================================================
2020-10-31
Saturday, October 31, 2020

Pushed two commits that import the RetroBSD repo into this project repo.
I also included the zip from the autobuilder so that there is some starting
point that is easily referenced (instead of just a commit id).


The last while I've been trying to get a mips-elf port of gcc, binutils,
and gdb to build the RetroBSD repo (which is mips-based). I had already got
it to work on Ubuntu (through the Windows Subsystem for Linux) by following
the instructions at http://retrobsd.org/wiki/doku.php/doc/toolchain-mips
which took quite a bit of time to go through.

Also, these instructions are made for an old version of gcc, compiled with
another old version of gcc. The newer gcc's are way more pedantic and find
so many more bugs, even small ones, so it has been a challenge. For
instance, one bug I came across was that glibc changed the location of some
defines into a different header, so I had to find out which one it was and
make a patch for it. It's not hard, just tedious and unfulfilling work.


Get first crack at a devel/mips-elf OpenBSD port done. It has gcc 4.8.1 and
binutils 2.27. Many patches were made. 8 for binutils, 19 for gcc.
I based this port on the ports devel/xtensa-elf and devel/arm-none-eabi.
gdb and newlib are not done yet (and maybe never will be).
binutils was the first I got working. It took a while to get gcc to work.

This port is for compiling the original code of RetroBSD on OpenBSD. The
original code targets the PIC32 MIPS-based microcontroller, thus mips-elf.

There are two or three segfaults in the process of building gcc. I have no
reason why, and I haven't investigated. Starting the build over again
allows a bit more progress until there is another segfault. I just kept
restarting it until the build is complete. The segfaults could be due to
running out of memory or something like that. Who knows. gcc is complex.

There are some post-patch and post-install fixes in the devel/avr/gcc
Makefile that need to be used. This will get rid of all the ridiculous
patches for the single space character needed by c++11 for variables.

===========================================================================
2020-11-01
Sunday, November 01, 2020

I started to build DiscoBSD with my new toolchain, but there are quite a
few errors. The main one is (after much investigation into the cause) due
to gcc not passing -msoft-float to the assembler. This caused the assembler
to assemble as -mhard-float. The linker wasn't able to link these two
separate types of objects.

I fixed this by getting a patchset from gcc, which I had to modify, as it
was meant for some newer version of gcc (5?).

I made 4 new patches, and added some more changes to an already existing
patch. That fixed the whole floating-point issue. The userland programs can
be built now. The kernel builds, too. There still is a problem with the
linker overlaying sections of the executable. I haven't researched it yet,
though.

The VirtualMIPS simulator isn't building. Probably due to not having the
libelf package on OpenBSD. I may need to port that library, too. Who knows.

Other problems are: elf2aout, manpages, init_main.c, expr, yacc kconfig
reduce/reduce conflicts, install not working.

I updated the README.md file for tools/openbsd to include how to get
binutils and gcc for the MIPS target.


Now I am trying to get VirtualMIPS to work. Most errors are about specific
types used. I am guessing they are defined in the libelf code.

I have had to (temporarily) hard-code some directory paths in the Makefile
for VirtualMIPS to what OpenBSD uses, such as /usr/local/include instead of
/opt/local/include on Linux.


Some cool things about the name DiscoBSD:
First, 'disco' is Latin for 'I learn' or 'to learn' or 'become aquainted
with' or 'I study, practice'. All of these definitions are quite in line
with the whole purpose of this directed study, which is to do a heck of a
lot of learning about Arm and about Unix and about early BSD.


sys/arch/pic32/dev/gpanel.c
/*
 * Generic TFT LCD driver for PIC32.
 * Supported chips: ST7781, NT35702, ILI9341.
 *
 * Copyright (C) 2015 Serge Vakulenko <serge@vak.ru>
 *
 * Permission to use, copy, modify, and distribute this software
 * and its documentation for any purpose and without fee is hereby
 * granted, provided that the above copyright notice appear in all
 * copies and that both that the copyright notice and this
 * permission notice and warranty disclaimer appear in supporting
 * documentation, and that the name of the author not be used in
 * advertising or publicity pertaining to distribution of the
 * software without specific, written prior permission.
 *
 * The author disclaim all warranties with regard to this
 * software, including all implied warranties of merchantability
 * and fitness.  In no event shall the author be liable for any
 * special, indirect or consequential damages or any damages
 * whatsoever resulting from loss of use, data or profits, whether
 * in an action of contract, negligence or other tortious action,
 * arising out of or in connection with the use or performance of
 * this software.
 */
#include <sys/param.h>
#include <sys/conf.h>
#include <sys/user.h>
#include <sys/ioctl.h>
#include <sys/systm.h>
#include <sys/uio.h>
#include <sys/tty.h>
#include <sys/config.h>
#include <sys/gpanel.h>

/*
 * Descriptor for access to the hardware-level driver.
 */
static struct gpanel_hw hw;

/*
 * Cursor position for text output.
 */
static int _col, _row;

/*
 * ID of the LCD controller chip.
 */
static int _chip_id;

/*
 * Display size.
 */
int gpanel_width, gpanel_height;

/*
 * Delay for 100 nanoseconds.
 * Needed to match the /WR and /RD timing requirements.
 */
#if CPU_KHZ <= 10000
#   define delay100ns()     /* empty */
#elif CPU_KHZ <= 20000
#   define delay100ns()     asm volatile("nop")
#elif CPU_KHZ <= 30000
#   define delay100ns()     asm volatile("nop; nop")
#elif CPU_KHZ <= 40000
#   define delay100ns()     asm volatile("nop; nop; nop")
#elif CPU_KHZ <= 50000
#   define delay100ns()     asm volatile("nop; nop; nop; nop")
#elif CPU_KHZ <= 60000
#   define delay100ns()     asm volatile("nop; nop; nop; nop; nop")
#elif CPU_KHZ <= 70000
#   define delay100ns()     asm volatile("nop; nop; nop; nop; nop; nop")
#else
#   define delay100ns()     asm volatile("nop; nop; nop; nop; nop; nop; nop; nop")
#endif

/*
 * Signal mappings:
 *  /RESET  - reset and initialize the chip.
 *  /CS     - chip select when low.
 *  /RD     - read operation enable.
 *  /WR     - write operation enable.
 *  RS      - command or data mode selection.
 *  D0-D7   - data bus, bidirectional.
 */
#define RST_IDLE()      LAT_SET(LCD_RST_PORT) = 1<<LCD_RST_PIN
#define RST_ACTIVE()    LAT_CLR(LCD_RST_PORT) = 1<<LCD_RST_PIN

#define CS_IDLE()       LAT_SET(LCD_CS_PORT) = 1<<LCD_CS_PIN
#define CS_ACTIVE()     LAT_CLR(LCD_CS_PORT) = 1<<LCD_CS_PIN

#define RD_IDLE()       LAT_SET(LCD_RD_PORT) = 1<<LCD_RD_PIN
#define RD_ACTIVE()     LAT_CLR(LCD_RD_PORT) = 1<<LCD_RD_PIN

#define WR_IDLE()       LAT_SET(LCD_WR_PORT) = 1<<LCD_WR_PIN
#define WR_ACTIVE()     LAT_CLR(LCD_WR_PORT) = 1<<LCD_WR_PIN

#define RS_DATA()       LAT_SET(LCD_RS_PORT) = 1<<LCD_RS_PIN
#define RS_COMMAND()    LAT_CLR(LCD_RS_PORT) = 1<<LCD_RS_PIN

/*
 * Set direction of data bus as output.
 */
void gpanel_write_dir()
{
    TRIS_CLR(LCD_D0_PORT) = 1 << LCD_D0_PIN;
    TRIS_CLR(LCD_D1_PORT) = 1 << LCD_D1_PIN;
    TRIS_CLR(LCD_D2_PORT) = 1 << LCD_D2_PIN;
    TRIS_CLR(LCD_D3_PORT) = 1 << LCD_D3_PIN;
    TRIS_CLR(LCD_D4_PORT) = 1 << LCD_D4_PIN;
    TRIS_CLR(LCD_D5_PORT) = 1 << LCD_D5_PIN;
    TRIS_CLR(LCD_D6_PORT) = 1 << LCD_D6_PIN;
    TRIS_CLR(LCD_D7_PORT) = 1 << LCD_D7_PIN;
}

/*
 * Set direction of data bus as input.
 */
void gpanel_read_dir()
{
    TRIS_SET(LCD_D0_PORT) = 1 << LCD_D0_PIN;
    TRIS_SET(LCD_D1_PORT) = 1 << LCD_D1_PIN;
    TRIS_SET(LCD_D2_PORT) = 1 << LCD_D2_PIN;
    TRIS_SET(LCD_D3_PORT) = 1 << LCD_D3_PIN;
    TRIS_SET(LCD_D4_PORT) = 1 << LCD_D4_PIN;
    TRIS_SET(LCD_D5_PORT) = 1 << LCD_D5_PIN;
    TRIS_SET(LCD_D6_PORT) = 1 << LCD_D6_PIN;
    TRIS_SET(LCD_D7_PORT) = 1 << LCD_D7_PIN;
}

/*
 * Control /CS signal.
 */
void gpanel_cs_active() { CS_ACTIVE(); }
void gpanel_cs_idle()   { CS_IDLE(); }

/*
 * Control /RS signal.
 */
void gpanel_rs_command() { RS_COMMAND(); }
void gpanel_rs_data()    { RS_DATA(); }

/*
 * Generate a /WR strobe.
 */
void gpanel_wr_strobe()
{
    delay100ns();
    WR_ACTIVE();
    delay100ns();
    WR_IDLE();
}

/*
 * Send a byte to the data bus.
 */
void gpanel_write_byte(int value)
{
    if (value & 1) {
        LAT_SET(LCD_D0_PORT) = 1 << LCD_D0_PIN;
    } else {
        LAT_CLR(LCD_D0_PORT) = 1 << LCD_D0_PIN;
    }
    if (value & 2) {
        LAT_SET(LCD_D1_PORT) = 1 << LCD_D1_PIN;
    } else {
        LAT_CLR(LCD_D1_PORT) = 1 << LCD_D1_PIN;
    }
    if (value & 4) {
        LAT_SET(LCD_D2_PORT) = 1 << LCD_D2_PIN;
    } else {
        LAT_CLR(LCD_D2_PORT) = 1 << LCD_D2_PIN;
    }
    if (value & 8) {
        LAT_SET(LCD_D3_PORT) = 1 << LCD_D3_PIN;
    } else {
        LAT_CLR(LCD_D3_PORT) = 1 << LCD_D3_PIN;
    }
    if (value & 0x10) {
        LAT_SET(LCD_D4_PORT) = 1 << LCD_D4_PIN;
    } else {
        LAT_CLR(LCD_D4_PORT) = 1 << LCD_D4_PIN;
    }
    if (value & 0x20) {
        LAT_SET(LCD_D5_PORT) = 1 << LCD_D5_PIN;
    } else {
        LAT_CLR(LCD_D5_PORT) = 1 << LCD_D5_PIN;
    }
    if (value & 0x40) {
        LAT_SET(LCD_D6_PORT) = 1 << LCD_D6_PIN;
    } else {
        LAT_CLR(LCD_D6_PORT) = 1 << LCD_D6_PIN;
    }
    if (value & 0x80) {
        LAT_SET(LCD_D7_PORT) = 1 << LCD_D7_PIN;
    } else {
        LAT_CLR(LCD_D7_PORT) = 1 << LCD_D7_PIN;
    }
    WR_ACTIVE();
    delay100ns();
    WR_IDLE();
}

/*
 * Read a byte from the data bus.
 */
int gpanel_read_byte()
{
    int value = 0;

    RD_ACTIVE();
    delay100ns();
    if (PORT_VAL(LCD_D0_PORT) & (1 << LCD_D0_PIN)) value |= 1;
    if (PORT_VAL(LCD_D1_PORT) & (1 << LCD_D1_PIN)) value |= 2;
    if (PORT_VAL(LCD_D2_PORT) & (1 << LCD_D2_PIN)) value |= 4;
    if (PORT_VAL(LCD_D3_PORT) & (1 << LCD_D3_PIN)) value |= 8;
    if (PORT_VAL(LCD_D4_PORT) & (1 << LCD_D4_PIN)) value |= 0x10;
    if (PORT_VAL(LCD_D5_PORT) & (1 << LCD_D5_PIN)) value |= 0x20;
    if (PORT_VAL(LCD_D6_PORT) & (1 << LCD_D6_PIN)) value |= 0x40;
    if (PORT_VAL(LCD_D7_PORT) & (1 << LCD_D7_PIN)) value |= 0x80;
    RD_IDLE();
    return value;
}

/*
 * Read a 16-bit value from the specified chip register.
 */
static int read_reg16(int reg)
{
    unsigned value;

    CS_ACTIVE();
    RS_COMMAND();
    gpanel_write_byte(reg >> 8);    /* Need 2 bytes of address for ST7781 */
    gpanel_write_byte(reg);
    gpanel_read_dir();              /* Switch data bus to input */
    RS_DATA();
    value = gpanel_read_byte() << 8;
    value |= gpanel_read_byte();
    gpanel_write_dir();             /* Restore data bus as output */
    CS_IDLE();
    return value;
}

/*
 * Read a 32-bit value from the specified chip register.
 */
static int read_reg32(int reg)
{
    unsigned value;

    CS_ACTIVE();
    RS_COMMAND();
    gpanel_write_byte(reg);
    gpanel_read_dir();              /* Switch data bus to input */
    RS_DATA();
    value = gpanel_read_byte() << 24;
    value |= gpanel_read_byte() << 16;
    value |= gpanel_read_byte() << 8;
    value |= gpanel_read_byte();
    gpanel_write_dir();             /* Restore data bus as output */
    CS_IDLE();
    return value;
}

/*
 * Draw a line.
 */
static void gpanel_draw_line(int x0, int y0, int x1, int y1, int color)
{
    int dx, dy, stepx, stepy, fraction;

    if (x0 == x1 || y0 == y1) {
        hw.fill_rectangle(x0, y0, x1, y1, color);
        return;
    }

    /* Use Bresenham's line algorithm. */
    dy = y1 - y0;
    if (dy < 0) {
        dy = -dy;
        stepy = -1;
    } else {
        stepy = 1;
    }
    dx = x1 - x0;
    if (dx < 0) {
        dx = -dx;
        stepx = -1;
    } else {
        stepx = 1;
    }
    dy <<= 1;                           /* dy is now 2*dy */
    dx <<= 1;                           /* dx is now 2*dx */
    hw.set_pixel(x0, y0, color);
    if (dx > dy) {
        fraction = dy - (dx >> 1);      /* same as 2*dy - dx */
        while (x0 != x1) {
            if (fraction >= 0) {
                y0 += stepy;
                fraction -= dx;         /* same as fraction -= 2*dx */
            }
            x0 += stepx;
            fraction += dy;             /* same as fraction -= 2*dy */
            hw.set_pixel(x0, y0, color);
        }
    } else {
        fraction = dx - (dy >> 1);
        while (y0 != y1) {
            if (fraction >= 0) {
                x0 += stepx;
                fraction -= dy;
            }
            y0 += stepy;
            fraction += dx;
            hw.set_pixel(x0, y0, color);
        }
    }
}

/*
 * Draw a rectangular frame.
 */
static void gpanel_draw_frame(int x0, int y0, int x1, int y1, int color)
{
    hw.fill_rectangle(x0, y0, x1, y0, color);
    hw.fill_rectangle(x0, y1, x1, y1, color);
    hw.fill_rectangle(x0, y0, x0, y1, color);
    hw.fill_rectangle(x1, y0, x1, y1, color);
}

/*
 * Draw a circle.
 */
static void gpanel_draw_circle(int x0, int y0, int radius, int color)
{
    int f = 1 - radius;
    int ddF_x = 0;
    int ddF_y = -2 * radius;
    int x = 0;
    int y = radius;

    hw.set_pixel(x0, y0 + radius, color);
    hw.set_pixel(x0, y0 - radius, color);
    hw.set_pixel(x0 + radius, y0, color);
    hw.set_pixel(x0 - radius, y0, color);
    while (x < y) {
        if (f >= 0) {
            y--;
            ddF_y += 2;
            f += ddF_y;
        }
        x++;
        ddF_x += 2;
        f += ddF_x + 1;
        hw.set_pixel(x0 + x, y0 + y, color);
        hw.set_pixel(x0 - x, y0 + y, color);
        hw.set_pixel(x0 + x, y0 - y, color);
        hw.set_pixel(x0 - x, y0 - y, color);
        hw.set_pixel(x0 + y, y0 + x, color);
        hw.set_pixel(x0 - y, y0 + x, color);
        hw.set_pixel(x0 + y, y0 - x, color);
        hw.set_pixel(x0 - y, y0 - x, color);
    }
}

/*
 * Draw a character from a specified font.
 */
static void gpanel_draw_char(const struct gpanel_font_t *font,
    int color, int background, int sym)
{
    unsigned cindex, width;
    const unsigned short *bits;

    switch (sym) {
    case '\n':      /* goto next line */
        _row += font->height;
        _col = 0;
        if (_row > gpanel_height - font->height)
            _row = 0;
        return;
    case '\r':      /* carriage return - go to begin of line */
        _col = 0;
        return;
    case '\t':      /* tab replaced by space */
        sym = ' ';
        break;
    }

    if (sym < font->firstchar || sym >= font->firstchar + font->size)
        sym = font->defaultchar;
    cindex = sym - font->firstchar;

    /* Get font bitmap depending on fixed pitch or not. */
    if (font->width) {
        /* Proportional font. */
        width = font->width[cindex];
    } else {
        /* Fixed width font. */
        width = font->maxwidth;
    }
    if (font->offset) {
        bits = font->bits + font->offset[cindex];
    } else {
        bits = font->bits + cindex * font->height;
    }

    /* Draw a character. */
    hw.draw_glyph(font, color, background, _col, _row, width, bits);
    _col += width;
}

/*
 * Draw a string of characters.
 * TODO: Decode UTF-8.
 */
static void gpanel_draw_text(const struct gpanel_font_t *font,
    int color, int background, int x, int y, const char *text)
{
    int sym;

    _col = x;
    _row = y;
    for (;;) {
        sym = *text++;
        if (! sym)
            break;

        gpanel_draw_char(font, color, background, sym);
    }
}

int gpanel_open(dev_t dev, int flag, int mode)
{
    if (minor(dev) != 0)
        return ENODEV;
    return 0;
}

int gpanel_close(dev_t dev, int flag, int mode)
{
    return 0;
}

int gpanel_read(dev_t dev, struct uio *uio, int flag)
{
    return ENODEV;
}

int gpanel_write(dev_t dev, struct uio *uio, int flag)
{
    return ENODEV;
}

/*
 * TODO: check whether user pointers are valid.
 */
int gpanel_ioctl(dev_t dev, register u_int cmd, caddr_t addr, int flag)
{
    switch (cmd) {
        /*
         * Clear the whole screen with a given color.
         */
        case GPANEL_CLEAR: {
            struct gpanel_clear_t *param = (struct gpanel_clear_t*) addr;

            if (hw.resize != 0)
                hw.resize(&hw, param->xsize, param->ysize);
            hw.fill_rectangle(0, 0, gpanel_width - 1, gpanel_height - 1, param->color);
            param->xsize = gpanel_width;
            param->ysize = gpanel_height;
            break;
        }

        /*
         * Draw a single pixel.
         */
        case GPANEL_PIXEL: {
            struct gpanel_pixel_t *param = (struct gpanel_pixel_t*) addr;

            hw.set_pixel(param->x, param->y, param->color);
            break;
        }

        /*
         * Draw a line.
         */
        case GPANEL_LINE: {
            struct gpanel_line_t *param = (struct gpanel_line_t*) addr;

            gpanel_draw_line(param->x0, param->y0, param->x1, param->y1, param->color);
            break;
        }

        /*
         * Draw a rectangle frame.
         */
        case GPANEL_RECT: {
            struct gpanel_rect_t *param = (struct gpanel_rect_t*) addr;

            gpanel_draw_frame(param->x0, param->y0, param->x1, param->y1, param->color);
            break;
        }

        /*
         * Fill a rectangle with color.
         */
        case GPANEL_FILL: {
            struct gpanel_rect_t *param = (struct gpanel_rect_t*) addr;

            hw.fill_rectangle(param->x0, param->y0, param->x1, param->y1, param->color);
            break;
        }

        /*
         * Draw a circle.
         */
        case GPANEL_CIRCLE: {
            struct gpanel_circle_t *param = (struct gpanel_circle_t*) addr;

            gpanel_draw_circle(param->x, param->y, param->radius, param->color);
            break;
        }

        /*
         * Fill a rectangular area with the user-supplied data.
         */
        case GPANEL_IMAGE: {
            struct gpanel_image_t *param = (struct gpanel_image_t*) addr;

            hw.draw_image(param->x, param->y, param->width, param->height,
                param->image);
            break;
        }

        /*
         * Draw a character.
         */
        case GPANEL_CHAR: {
            struct gpanel_char_t *param = (struct gpanel_char_t*) addr;

            _col = param->x;
            _row = param->y;
            gpanel_draw_char(param->font, param->color, param->background, param->sym);
            break;
        }

        /*
         * Draw a string of characters.
         */
        case GPANEL_TEXT: {
            struct gpanel_text_t *param = (struct gpanel_text_t*) addr;

            gpanel_draw_text(param->font, param->color, param->background,
                param->x, param->y, param->text);
            break;
        }
    }
    return 0;
}

/*
 * Draw a BSD logo on the screen.
 */
static void draw_logo()
{
#define K       7
#define COLOR_B 0xf81f
#define COLOR_S 0x07ff
#define COLOR_D 0xffe0

    int x = gpanel_width/2  - 17*K;
    int y = gpanel_height/2 + 11*K;

    hw.fill_rectangle(0, 0, gpanel_width - 1, gpanel_height - 1, 0);

    /* B */
    gpanel_draw_line( 0*K+x, y- 0*K,  0*K+x, y-11*K, COLOR_B);
    gpanel_draw_line( 0*K+x, y-11*K,  0*K+x, y-22*K, COLOR_B);
    gpanel_draw_line( 0*K+x, y-22*K, 10*K+x, y-19*K, COLOR_B);
    gpanel_draw_line(10*K+x, y-19*K,  0*K+x, y-11*K, COLOR_B);
    gpanel_draw_line( 0*K+x, y-11*K, 10*K+x, y- 8*K, COLOR_B);
    gpanel_draw_line(10*K+x, y- 8*K,  0*K+x, y- 0*K, COLOR_B);

    /* S */
    gpanel_draw_line(22*K+x, y-22*K, 12*K+x, y-19*K, COLOR_S);
    gpanel_draw_line(12*K+x, y-19*K, 22*K+x, y- 8*K, COLOR_S);
    gpanel_draw_line(22*K+x, y- 8*K, 12*K+x, y- 0*K, COLOR_S);

    /* D */
    gpanel_draw_line(24*K+x, y-22*K, 24*K+x, y- 0*K, COLOR_D);
    gpanel_draw_line(24*K+x, y-22*K, 34*K+x, y-19*K, COLOR_D);
    gpanel_draw_line(34*K+x, y-19*K, 34*K+x, y- 8*K, COLOR_D);
    gpanel_draw_line(34*K+x, y- 8*K, 24*K+x, y- 0*K, COLOR_D);
}

/*
 * Read the the chip ID register.
 * Some controllers have a register #0
 * programmed with unique chip ID.
 */
static int read_id()
{
    int id, retry;

    /* Some controllers have a register #0
     * programmed with unique chip ID. */
    id = read_reg16(0);
    if (id != 0)
        return id;

    /* Try ID from register #4. */
    id = read_reg32(4) & 0xffffff;
    if (id != 0)
        return id;

    /* Try ID from register #D3.
     * Might need to wait until the register becomes alive after Reset. */
    for (retry=0; retry<5; retry++) {
        id = read_reg32(0xD3) & 0xffffff;
        if (id != 0)
            return id;
        udelay(50000);
    }
    return 0;
}

/*
 * Detect the type of the LCD controller, and initialize it.
 * Return true if found and initialized ok.
 */
static int probe(config)
    struct conf_device *config;
{
    /*
     * Set all control bits to high (idle).
     * Signals are active low.
     */
    CS_IDLE();
    WR_IDLE();
    RD_IDLE();
    RST_IDLE();

    /* Enable outputs. */
    TRIS_CLR(LCD_CS_PORT) = 1 << LCD_CS_PIN;
    TRIS_CLR(LCD_RS_PORT) = 1 << LCD_RS_PIN;
    TRIS_CLR(LCD_WR_PORT) = 1 << LCD_WR_PIN;
    TRIS_CLR(LCD_RD_PORT) = 1 << LCD_RD_PIN;
    TRIS_CLR(LCD_RST_PORT) = 1 << LCD_RST_PIN;
    gpanel_write_dir();

    /* Reset the chip. */
    RST_ACTIVE();
    udelay(1000);
    RST_IDLE();
    udelay(1000);

    /* Read the the chip ID register.
     * Some controllers have a register #0
     * programmed with unique chip ID. */
    _chip_id = read_id();
    switch (_chip_id) {
    default:
        printf("gpanel0: Unknown chip ID = 0x%04x\n", _chip_id);
        goto failed;

    case 0x7783:
        /* Sitronix ST7781. */
        st7781_init_display(&hw);
        break;

    case 0x009341:
        /* Ilitek ILI9341. */
        ili9341_init_display(&hw);
        break;

    case 0x012200:
        /* Ilitek ILI9481. */
        ili9481_init_display(&hw);
        break;

    case 0x388000:
        /* Novatek NT35702. */
        nt35702_init_display(&hw);
        break;
    }
    printf("gpanel0: <%s> display %ux%u\n", hw.name, gpanel_width, gpanel_height);
    draw_logo();
    return 1;

failed:
    /* Disable outputs. */
    gpanel_read_dir();
    TRIS_SET(LCD_CS_PORT) = 1 << LCD_CS_PIN;
    TRIS_SET(LCD_RS_PORT) = 1 << LCD_RS_PIN;
    TRIS_SET(LCD_WR_PORT) = 1 << LCD_WR_PIN;
    TRIS_SET(LCD_RD_PORT) = 1 << LCD_RD_PIN;
    TRIS_SET(LCD_RST_PORT) = 1 << LCD_RST_PIN;
    return 0;
}

struct driver gpaneldriver = {
    "gpanel", probe,
};
